recommender/engine/rl_agent/preprocessing/state_embedder.py

```diff
@@ -1,45 +1,136 @@
+            self.user_embedder = user_embedder
+        else:
+            self.user_embedder = load_last_module(USERS_AUTOENCODER).encoder
+
+        if service_embedder is not None:
+            self.service_embedder = service_embedder
+        else:
+            self.service_embedder = load_last_module(SERVICES_AUTOENCODER).encoder
 
-        # TODO: implement the rest of initialization (lazy embedders loading)
-        pass
+        if search_phrase_embedder is not None:
+            self.search_phrase_embedder = search_phrase_embedder
+        else:
+            self.search_phrase_embedder = SearchPhraseEmbedder()
+
+        if filters_embedder is not None:
+            self.filters_embedder = filters_embedder
+        else:
+            self.filters_embedder = FiltersEmbedder()
 
     def __call__(self, state: State) -> Tuple[torch.Tensor]:
         """
         Transform state to the tuple of tensors using appropriate embedders as
          follows:
-            - state.user -> one_hot_tensor --(user_embedder)--> user_tensor
-            - state.user.accessed_services and clicks information??? -> services_one_hot_tensors --(service_embedder)--> services_dense_tensors --(concat)--> services_history_tensor
-            - state.last_search_data...filters(multiple fields) -> one_hot_tensor --(filters_embedder)--> filters_tensor
-            - state.last_search_data.q --(searchphrase_embedder)--> search_phrase_tensor
+            - state.user
+                -> one_hot_tensor
+                 --(user_embedder)--> user_tensor
+            - state.services_history
+                -> services_one_hot_tensors
+                --(service_embedder)--> services_dense_tensors
+                --(concat)--> services_history_tensor
+            - state.last_search_data.q
+                --(searchphrase_embedder)--> search_phrase_tensor
+            - state.last_search_data...filters(multiple fields)
+                -> one_hot_tensor
+                --(filters_embedder)--> filters_tensor
 
         Args:
-            state: State object that contains information about user, user's services history and searchphrase.
+            state: State object that contains information about user, user's
+             services history and searchphrase.
 
         Returns:
             Tuple of following tensors (in this order):
-                - user_tensor of shape [UE]
-                - services_history_tensor of shape [N, SE]
-                - filters_tensor of shape [FE]
-                - search_phrase_tensor of shape [SPE]
+                - user of shape [UE]
+                - services_history of shape [N, SE]
+                - search_phrase of shape [X, SPE]
+                - filters of shape [SE]
+
                 where:
                     - UE is user content tensor embedding dim
-                    - N is user clicked services history length
+                    - N is user clicked services history length - it is not constant
```

Comment:
maybe it should be lowercase "n" if it's not constant? - same for X

Reply:
If it was a variable/constant in the code - yes, but it is just dimension in the docstring, so it has its own convention.